### PR TITLE
Modify/read marker2

### LINKED ISSUE #507: Race condition with read marker
Current ReadMarker structure:
```
type ReadMarker struct {
	ClientID        string           `json:"client_id"`
	ClientPublicKey string           `json:"client_public_key"`
	BlobberID       string           `json:"blobber_id"`
	AllocationID    string           `json:"allocation_id"`
	OwnerID         string           `json:"owner_id"`
	Timestamp       common.Timestamp `json:"timestamp"`
	ReadCounter     int64            `json:"counter"`
	Signature       string           `json:"signature"`
}
```

As per the protocol _**The smart contract will release funds corresponding to the difference between the read_counter in the marker and the latest read_counter redeemed on the blockchain**_

The importance of ReadCounter is only to allow blobber to redeem tokens from read pools. The ReadCounter is based on the assumption that each counter corresponds to 64KB read by the client. So the price calculation is done as: `ΔReadcounter * blockSize * read_price`. 

_Note: BlockSize is customizable with new chunkupload feature so it is no more fixed 64KB._

However above implementation leads to race condition for ReadCounter as readmarker signed by client requires it to be the latest ReadCounter so that blobber can calculate price easily. Blobber code rejects download request if readCounter is not in sync.
```
	if latestRM != nil &&
		latestRM.ReadCounter+(numBlocks) != readMarker.ReadCounter {

		var response = &blobberhttp.DownloadResponse{
			Success:      false,
			LatestRM:     latestRM,
			Path:         fileref.Path,
			AllocationID: fileref.AllocationID,
		}
		return response, nil
	}
```

ReadMarker is associated to client so there won't be race between clients for ReadCounter as they will have their own ReadCounter. However if a client sends multiple download requests, be it for same file or different files, there will be race for latest ReadCounter between requests.

In my view this problem can be solved by replacing ReadCounter with ReadSize and requiring blobber to request read pools with each read marker to redeem read tokens rather than single read marker being able to redeem whole tokens. With this client simply signs readmarker with required size and there is no race for anything.

EDIT:
After successfully merging the above solution, we realized that there is no mechanism against indefinite redeeming of same read marker. `ReadCounter` had served this purpose. So we moved

### PR DESCRIPTION
**Pull request by lpoli**
This PR contains the following changes:

    Change ReadMarker's ReadCounter to ReadSize
    Remove latest ReadMarker logic from the code as it is useless with current implementation
    Modify pendings table to accommodate above changes
    Reduce unnecessary functions and function calls in worker in readmarker package

This PR solves the issue https://github.com/0chain/blobber/issues/507

### CODE REVIEW

**Comment by service-0chain** [comment]
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | modify/read-marker  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | modify/read-marker  |
| zwalletcli  | staging  |
| 0chain  | modify/read-marker  |
| blobber  | modify/read-marker2  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | staging  |
| 0proxy  | staging  |
| 0search  | staging  |

**Comment by service-0chain** [comment]
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | master  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | modify/read-marker  |
| zwalletcli  | staging  |
| 0chain  | modify/read-marker  |
| blobber  | modify/read-marker2  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | staging  |
| 0proxy  | staging  |
| 0search  | staging  |

File: code/go/0chain.net/blobbercore/allocation/entity.go [hunk 1]
```diff
@@ -101,102 +106,104 @@ func (a *Allocation) WantWrite(blobberID string, size int64, wmt common.Timestam
...
 	p = new(Pending)
-	err = tx.Model(&Pending{}).
-		Where(query, clientID, allocationID, blobberID).
-		First(&p).Error
+	err = tx.Model(&Pending{}).Where("id=?", clientID+":"+allocationID).First(p).Error
 	if errors.Is(err, gorm.ErrRecordNotFound) {
-		p.ClientID = clientID
-		p.AllocationID = allocationID
-		p.BlobberID = blobberID
 		err = tx.Create(p).Error
```
Comment (peterlimg) [inline code/go/0chain.net/blobbercore/allocation/entity.go]: The p is empty here, is this an intended behavior or a mistake?

File: code/go/0chain.net/blobbercore/allocation/entity.go [hunk 2]
```diff
@@ -248,15 +254,43 @@ func (*WritePool) TableName() string {
...
+	}
+	return *b, nil
+}
+
 func SetReadPools(db *gorm.DB, clientID, allocationID, blobberID string, rps []*ReadPool) (err error) {
 	// cleanup and batch insert (remove old pools, add / update new)
 	const query = `client_id = ? AND
-        allocation_id = ? AND
-        blobber_id = ?`
+			allocation_id = ?`
```
Comment (peterlimg) [inline code/go/0chain.net/blobbercore/allocation/entity.go:260]: If the blobber_id param is not being used, please remove it.

File: code/go/0chain.net/blobbercore/allocation/entity.go [hunk 3]
```diff
@@ -275,8 +309,8 @@ func SetReadPools(db *gorm.DB, clientID, allocationID, blobberID string, rps []*
 
 func SetWritePools(db *gorm.DB, clientID, allocationID, blobberID string, wps []*WritePool) (err error) {
 	const query = `client_id = ? AND
-        allocation_id = ? AND
-        blobber_id = ?`
+				allocation_id = ? AND
+				blobber_id = ?`
```
Comment (peterlimg) [inline code/go/0chain.net/blobbercore/allocation/entity.go]: Does WritePool have the blobber_id

**Review by peterlimg — CHANGES_REQUESTED** [review]

File: code/go/0chain.net/blobbercore/allocation/entity.go (hunk 1 above)
Comment (lpoli) [reply code/go/0chain.net/blobbercore/allocation/entity.go]: This function should have been deleted. Forgot to delete.
Will fix this

File: code/go/0chain.net/blobbercore/allocation/entity.go (hunk 3 above)
Comment (lpoli) [reply code/go/0chain.net/blobbercore/allocation/entity.go]: No it does not.
Fixed now. Thank you.

**Review by peterlimg — APPROVED** [review]
lgtm

**Comment by service-0chain** [comment]
Manual system tests [success] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | fix/download-cost-calculation  |
| gosdk  | modify/read-marker  |
| zboxcli  | modify/read-marker  |
| zwalletcli  | staging  |
| 0chain  | modify/read-marker  |
| blobber  | modify/read-marker2  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | staging  |
| 0proxy  | staging  |
| 0search  | staging  |

**Pull request merged**